engine/updategraph/src/main/java/io/deephaven/engine/updategraph/UpdateGraphAwareCompletableFuture.java

```diff
@@ -0,0 +1,164 @@
+import org.jetbrains.annotations.Nullable;
+
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.concurrent.locks.Condition;
+
+public class UpdateGraphAwareCompletableFuture<T> implements Future<T> {
+
+    private final UpdateGraph updateGraph;
+
+    /** This condition is used to signal any threads waiting on the UpdateGraph exclusive lock. */
+    private volatile Condition updateGraphCondition;
+
+    private final FunctionalLock lock = new FunctionalReentrantLock();
+    private volatile Condition lockCondition;
+
+    private volatile ThrowingSupplier<T, ExecutionException> resultSupplier;
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<UpdateGraphAwareCompletableFuture, ThrowingSupplier> RESULT_UPDATER =
+            AtomicReferenceFieldUpdater.newUpdater(
+                    UpdateGraphAwareCompletableFuture.class, ThrowingSupplier.class, "resultSupplier");
+
+    /** The encoding of the cancelled supplier. */
+    private static final ThrowingSupplier<?, ExecutionException> CANCELLATION_SUPPLIER = () -> {
+        throw new CancellationException();
+    };
+
+    public UpdateGraphAwareCompletableFuture(@NotNull final UpdateGraph updateGraph) {
+        this.updateGraph = updateGraph;
+    }
+
+    ////////////////
+    // Future API //
+    ////////////////
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+        // noinspection unchecked
+        return trySignalCompletion((ThrowingSupplier<T, ExecutionException>) CANCELLATION_SUPPLIER);
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return resultSupplier == CANCELLATION_SUPPLIER;
+    }
+
+    @Override
+    public boolean isDone() {
+        return resultSupplier != null;
+    }
+
+    @Override
+    public T get() throws InterruptedException, ExecutionException {
+        if (resultSupplier != null) {
+            return resultSupplier.get();
+        }
+        try {
+            return getInternal(0, null);
+        } catch (TimeoutException toe) {
+            throw new IllegalStateException("Unexpected TimeoutException", toe);
+        }
+    }
+
+    @Override
+    public T get(final long timeout, @NotNull final TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        if (resultSupplier != null) {
+            return resultSupplier.get();
+        }
+        if (timeout <= 0) {
+            throw new TimeoutException();
+        }
+        return getInternal(timeout, unit);
+    }
+
+    private T getInternal(long timeout, @Nullable TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        // test lock conditions
+        if (updateGraph.sharedLock().isHeldByCurrentThread()) {
+            throw new UnsupportedOperationException(
+                    "Cannot Future#get while holding the " + updateGraph + " shared lock");
+        }
```

Comment:
Put this in a method and call it first thing in both `get` overloads. We want to be consistent in our error delivery.
```suggestion
        if (updateGraph.sharedLock().isHeldByCurrentThread()) {
            throw new UnsupportedOperationException(
                    "Cannot Future.get(...) while holding the " + updateGraph + " shared lock");
        }
```